Add support for EPA

paraddise opened the pull request:
TODO:
- [x] Implement for NTLM tds < 8.0
- [x] Implement for WinSSPI
- [ ] Implement for Kerberos tds < 8.0
- [ ] Implement for tds 8.0

paraddise commented:
As I understand I need to pass tls-unique to `gss_init_sec_context ` as a field in `gss_channel_bindings_struct`, and it seems I need to patch https://github.com/jcmturner/gokrb5

shueybubbles commented on msdsn/conn_str.go:
```diff
@@ -569,6 +571,14 @@ func Parse(dsn string) (Config, error) {
 	}
 
+	disableEPA, ok := params[DisableEPA]
+	if ok {
```
>if ok { [](http://example.com/codeflow?start=1&length=7)

perhaps for the first release the default behavior should be the same as before and only use the EPA code if `disableEPA` is added to the connection string explicitly as `false` or `0`. It could also use an env var like `MSSQL_USE_EPA` to control the behavior for folks that don't want to have to update their connection strings.

shueybubbles commented on msdsn/conn_str.go:
```diff
@@ -639,6 +643,19 @@ func Parse(dsn string) (Config, error) {
+	if !ok {
+		epaString = os.Getenv("MSSQL_USE_EPA")
+	}
+	switch strings.ToLower(epaString) {
```
>switch strings.ToLower(epaString) [](http://example.com/codeflow?start=1&length=33)

`strconv.ParseBool` ?

paraddise commented:
1. removed EpaMode string.
2. There is a problem that we don't know how to calculate CB for tls 1.3 when connecting to mssql. Calculating CB via TLS_EXPORTER_EKM_LABEL like in RFC doesn't work and microsoft has no information how calculate channel binding for tls 1.3. Maybe someone will find the way how it's calculated but now there is no working examples for mssql in opensource. Microsoft tools work with tls1.3 maybe microsoft will share this information in the future and I think we can keep related code for future use
3. you are right, I hadn't thought about it. Fixed

paraddise commented:
To support EPA for krb5 I need to pass channel_binding structure to InitializeContext call, but for that I need corresponding parameters in this function, for that I need to make PR to krb5 repo
